### PR TITLE
fix(mission_planner): routing api uninitialization

### PR DESCRIPTION
**Pull request by TakaHoribe**
## Description

To fix the issue: "the set_route api is rejected even though the autoware_state is `WAIT_FOR_ROUTE`".

Publish `routing_state = UNKNOWN` until the routing system is not initialized.

Without this PR, it gets `autoware_state = WAIT_FOR_ROUTE` and `routing_state = UNSET` even if the mission_planner is not ready. Due to this inconsistency, calling set_route while in the wait_for_route state causes the API to fail.

Before this PR, when the ego pose is not set, the route state is UNSET:
```
ros2 topic echo /planning/mission_planning/route_state
stamp:
  sec: 1694341968
  nanosec: 326244268
state: 0
```

After this PR, when the ego pose is not set, the route state is UNKNOWN:
```
ros2 topic echo /planning/mission_planning/route_state
stamp:
  sec: 1694342360
  nanosec: 163490802
state: 1
---
```

## Related links

[TIERIV Internal ticket](https://tier4.atlassian.net/browse/RT1-3637)

## Tests performed

Run psim

## Notes for reviewers

check 

## Interface changes

None

## Effects on system behavior

route_state behavior will change

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by TakaHoribe** [comment]
@kosuke55 @isamu-takagi This is to fix a specific issue, but I'm not sure this fix is the best way. Would you check this.

**Comment by isamu-takagi** [comment]
UNKNOWN is reserved as an invalid value, so please avoid publishing this as a normal value.

**Comment by TakaHoribe** [comment]
@isamu-takagi 
> UNKNOWN is reserved as an invalid value, so please avoid publishing this as a normal value.

Thank you for your comment. Do you have an idea how to tell the routing system is not initialized yet? Should we add the new `UNINITIALIZED` field in the message?

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 1]
```diff
@@ -301,18 +329,15 @@ void MissionPlanner::on_set_route(
...
 
-  if (state_.state != RouteState::Message::UNSET) {
-    throw component_interface_utils::ServiceException(
-      ResponseCode::ERROR_ROUTE_EXISTS, "The route is already set.");
-  }
-  if (!planner_->ready()) {
+  if (state_.state == RouteState::Message::UNKNOWN) {  // should be UNINITIALIZED
     throw component_interface_utils::ServiceException(
-      ResponseCode::ERROR_PLANNER_UNREADY, "The planner is not ready.");
+      ResponseCode::ERROR_ROUTE_EXISTS, "The planner is not ready.");
```
Comment (isamu-takagi) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:334]:
```suggestion
      ResponseCode::ERROR_PLANNER_UNREADY, "The planner is not ready.");
```

File: planning/mission_planner/src/mission_planner/mission_planner.cpp [hunk 2]
```diff
@@ -574,17 +586,13 @@ void MissionPlanner::on_change_route(
 {
   using ResponseCode = autoware_adapi_v1_msgs::srv::SetRoute::Response;
 
-  if (state_.state != RouteState::Message::SET) {
+  if (state_.state == RouteState::Message::UNKNOWN) {  // should be UNINITIALIZED
     throw component_interface_utils::ServiceException(
-      ResponseCode::ERROR_INVALID_STATE, "The route hasn't set yet. Cannot reroute.");
+      ResponseCode::ERROR_ROUTE_EXISTS, "The planner is not ready.");
```
Comment (isamu-takagi) [inline planning/mission_planner/src/mission_planner/mission_planner.cpp:591]:
```suggestion
      ResponseCode::ERROR_PLANNER_UNREADY, "The planner is not ready.");
```

**Comment by isamu-takagi** [comment]
If you want to know whether a route request is possible, I think you should add a flag to indicate this. This is because when considering reroutes and MRM, it is easier to handle them separately from states.

**Comment by TakaHoribe** [comment]
@isamu-takagi 
>If you want to know whether a route request is possible, I think you should add a flag to indicate this. This is because when considering reroutes and MRM, it is easier to handle them separately from states.

I feel this makes it more confusing. The issue we want to fix now is, "the set_route API is rejected even though the autowre_state is WAIT_FOR_ROUTE".

Now the autoware_state is deprecated, but if the role of the `routing_state` is to tell the status of the routing system, it should contain the uninitialized information.


> it is easier to handle them separately from states.
MRM system can see the status through the response of the API, so it does not matter here.

**Comment by TakaHoribe** [comment]
Note:

Even if the `UNINITIALIZED` status is supported, sometimes the set_route might be rejected due to the planning status.

The status of whether the module can receive the route should be managed at a higher level, not only in the mission_planner.

@isamu-takagi will keep investigating the better way. (I'll make an issue to track this.)

**Comment by TakaHoribe** [comment]
>To fix the issue: "the set_route api is rejected even though the autoware_state is WAIT_FOR_ROUTE".

I changed the way to fix the issue. Now it does not publish the route_state until the node is ready for the API.

**Review by kosuke55 — APPROVED** [review]

**Review by isamu-takagi — APPROVED** [review]
@TakaHoribe 
LGTM. ~However, it need to be fixed so it doesn't wait in the constructor in the future. Because it blocks the initialization of other components.~

I had a misunderstanding. No problem.

**Comment by TakaHoribe** [comment]
@isamu-takagi Thank you for your review. It is already no-wait in the constructor. The time starts running after the constructor is completed, and the route_state msg will be published when all conditions checked in the timer are satisfied.

**Pull request merged**